nixos/modules/services/network-filesystems/samba.nix

```diff
@@ -3,35 +3,12 @@
 with lib;
 
 let
-
-  smbToString = x: if builtins.typeOf x == "bool"
-                   then boolToString x
-                   else toString x;
-
   cfg = config.services.samba;
 
-  samba = cfg.package;
-
-  shareConfig = name:
-    let share = getAttr name cfg.shares; in
-    "[${name}]\n " + (smbToString (
-       map
-         (key: "${key} = ${smbToString (getAttr key share)}\n")
-         (attrNames share)
-    ));
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "smb.conf" cfg.settings;
 
-  configFile = pkgs.writeText "smb.conf"
-    (if cfg.configText != null then cfg.configText else
-    ''
-      [global]
-      security = ${cfg.securityType}
-      passwd program = /run/wrappers/bin/passwd %u
-      invalid users = ${smbToString cfg.invalidUsers}
```

Comment:
if the user follows the instruction to use `services.samba.settings`, these 2 options are lost without an instruction on how to migrate them.

Reply:
@yshui I think `invalid users` and `securtiy` are renamed. The only option thet is indeed missing is the `passwd program`.

Reply:
@bachp right, but `invalidUsers` had a default value, so if the user didn't config anything they would still get a `invalid users = root` by default. after this migration that's gone, and the user won't get any warnings about that.

Reply:
@yshui Thank you for your feedback! :)
Yes you're right, I shouldn't break the default values like I did here.

What do you think if I put them back in the submodule options?

```nix
       settings = lib.mkOption {
-        type = lib.types.submodule { freeformType = settingsFormat.type; };
-        default = {};
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
+          options = {
+            "global"."security" = lib.mkOption {
+              type = lib.types.enum [ "auto" "user" "domain" "ads" ];
+              default = "user";
+              description = "Samba security type.";
+            };
+            "global"."invalid users" = lib.mkOption {
+              type = lib.types.str;
+              default = "root";
+              description = "List of users who are denied to login via Samba.";
+            };
+            "global"."passwd program" = lib.mkOption {
+              type = lib.types.str;
+              default = "/run/wrappers/bin/passwd %u";
+              description = "FIXME";
+            };
+          };
+        };
+        default = {
+          "global" = {
+            "security" = "user";
+            "passwd program" = "/run/wrappers/bin/passwd %u";
+            "invalid users" = "root";
+          };
+        };
         example = {
           "global" = {
             "security" = "user";
```

Reply:
https://github.com/NixOS/nixpkgs/pull/340500